src/Compiler/Microsoft.AspNetCore.Razor.Language/test/IntegrationTests/ComponentCodeGenerationTestBase.cs

```diff
@@ -6827,6 +6827,191 @@ public class Column<TItem> : ComponentBase
+
+            namespace Test;
+
+            [CascadingTypeParameter(nameof(T))]
+            public class Grid<T> : ComponentBase
+            {
+                [Parameter] public Func<Dictionary<X, T>>? Data { get; set; }
+            }
+
+            public partial class GridColumn<T> : ComponentBase
+            {
+            }
+            
+            public class X { }
+            """));
+
+        // Act
+        var generated = CompileToCSharp("""
+            <Grid Data="@(() => new Dictionary<X, string>())">
+                <GridColumn />
+            </Grid>
+            """, nullableEnable: true);
+
+        // Assert
+        AssertDocumentNodeMatchesBaseline(generated.CodeDocument);
+        AssertCSharpDocumentMatchesBaseline(generated.CodeDocument);
+        CompileToAssembly(generated);
+    }
+
+    [IntegrationTestFact, WorkItem("https://github.com/dotnet/razor/issues/9631")]
+    public void CascadingGenericInference_GenericArgumentNested_Dictionary_03()
+    {
+        // Arrange
+        AdditionalSyntaxTrees.Add(Parse("""
+            using Microsoft.AspNetCore.Components;
+            using System;
+            using System.Collections.Generic;
+
+            namespace Test
+            {
+                [CascadingTypeParameter(nameof(T))]
+                public class Grid<T> : ComponentBase
+                {
+                    [Parameter] public Dictionary<X, T>? Data { get; set; }
+                }
+
+                public partial class GridColumn<T> : ComponentBase
+                {
+                }
+            }
+
+            public class X { }
+            """));
+
+        // Act
+        var generated = CompileToCSharp("""
+            <Grid Data="@(new Dictionary<X, string>())">
+                <GridColumn />
+            </Grid>
+            """, nullableEnable: true);
+
+        // Assert
+        AssertDocumentNodeMatchesBaseline(generated.CodeDocument);
+        AssertCSharpDocumentMatchesBaseline(generated.CodeDocument);
+        CompileToAssembly(generated);
+    }
+
+    [IntegrationTestFact, WorkItem("https://github.com/dotnet/razor/issues/9631")]
+    public void CascadingGenericInference_GenericArgumentNested_Dictionary_Dynamic()
+    {
+        // Arrange
+        AdditionalSyntaxTrees.Add(Parse("""
+            using Microsoft.AspNetCore.Components;
+            using System;
+            using System.Collections.Generic;
+
+            namespace Test
+            {
+                [CascadingTypeParameter(nameof(T))]
+                public class Grid<T> : ComponentBase
+                {
+                    [Parameter] public Dictionary<dynamic, T>? Data { get; set; }
+                }
+
+                public partial class GridColumn<T> : ComponentBase
+                {
+                }
+            }
+            """));
+
+        // Act
+        var generated = CompileToCSharp("""
+            <Grid Data="@(new Dictionary<dynamic, string>())">
```

Comment:
Note that `dynamic` still doesn't work unlike everything else I can think of. That's because every other type argument here is `global::` prefixed (e.g., `object` becomes `global::System.Object`), except `dynamic` which stays the same, hence the generic type parameter parser thinks it's a type parameter as well. Maybe we could fix that by special-casing `dynamic`.